Implement name soft deletes

clemens-msupply opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #3555

# 👩🏻‍💻 What does this PR do?
When receiving a name MERGE record, the deleted name gets soft-delete now.

# 🧪 Testing

See issue, but found this easier:
- add a dispensary store to the central server site
- create name "a1" and "a2" on the central server
- sync and see names on the remote sites
- merge "a1" and "a2"
- sync, there should only be one name visible on the remote site now

clemens-msupply commented on package.json:
```diff
@@ -1,7 +1,7 @@
   "name": "open-msupply",
   "//": "Main version for the app, should be in semantic version format (any release candidate or test build should be separated by '-' i.e. 1.1.1-rc1 or 1.1.1-test",
-  "version": "2.0.00",
+  "version": "2.1.00",
```
Could sqeeze it in 2.0.0?

clemens-msupply commented on server/repository/src/db_diesel/name_row.rs:
```diff
@@ -212,8 +217,11 @@ impl<'a> NameRowRepository<'a> {
     }
 
+    /// Soft delete a name record (set is_active to false)
     pub fn delete(&self, name_id: &str) -> Result<(), RepositoryError> {
```
Wonder if we should rename these soft delete method to `soft_delete()`

roxy-dao replied:
Yeah, would be nice to rename to match what the function actually does

clemens-msupply commented on server/service/src/sync/translations/mod.rs:
```diff
@@ -336,6 +338,7 @@ pub(crate) trait SyncTranslation {
         Ok(PullTranslateResult::NotMatched)
     }
 
+    /// Translate a delete record received from the central server(s)
```
Got confused about from/to and added some comments...

clemens-msupply commented on server/service/src/sync/translations/name.rs:
```diff
@@ -280,12 +281,18 @@ impl SyncTranslation for NameTranslation {
+        if !is_active {
+            return Ok(PushTranslateResult::Ignored(
+                "Ignore pushing soft deleted name".to_string(),
+            ));
```
It got deleted in mSupply and we soft-deleted it. So I guess it correct to not push it back...

roxy-dao commented on server/service/src/sync/translations/name.rs:
```diff
@@ -230,12 +231,12 @@ impl SyncTranslation for NameTranslation {
                 .or(created_date.map(|date| date.and_hms_opt(0, 0, 0).unwrap())),
             date_of_death,
             custom_data_string,
+            is_active: existing_name.map_or(true, |name| name.is_active),
```
OG doesn't have an `is_active` field, so this will always be true.

It does have an `is_deleted` field, but that doesn't actually get used so any merged names are deleted...

clemens-msupply replied:
Don't think its possible right now, but thought about some case where the name is soft-deleted locally and another UPSERT record is received...

However, as of https://github.com/msupply-foundation/open-msupply/pull/3693 I plan to change it to `deleted_datetime: Option<_>` which will solve it nicely as well.

roxy-dao replied:
The issue is that the merged name is still viewable for me tho? And the is_active bool is true for the merged name

roxy-dao reviewed (changes requested):
Nice! Just a few notes down below (: I think maybe we might be able to squeeze into 2.0.0?

clemens-msupply commented:
> Nice! Just a few notes down below (: I think maybe we might be able to squeeze into 2.0.0?

Done, also renamed `mark_deleted()` and changed to `created_datetime` column

roxy-dao commented on server/service/src/sync/translations/name.rs:
```diff
@@ -230,12 +231,12 @@ impl SyncTranslation for NameTranslation {
                 .or(created_date.map(|date| date.and_hms_opt(0, 0, 0).unwrap())),
             date_of_death,
             custom_data_string,
+            deleted_datetime: existing_name.and_then(|name| name.deleted_datetime),
```
Why is this using existing_name? If the name was merged in OG, then this wouldn't populate the deleted datetime at all....

roxy-dao replied:
and we're not deleting patients in OMS....

clemens-msupply replied:
Yeah you are right it's not needed.

roxy-dao reviewed (approved):
Thanks for the changes Clemens~
